v2/internal/controllers/crd_networking_applicationgateway_test.go

```diff
@@ -0,0 +1,389 @@
+		RuleSetType:    to.Ptr[string]("OWASP"),
+		RuleSetVersion: to.Ptr[string]("3.2"),
+	})
+	armid_params := map[string]string{
+		"param-key1": "applicationGateways",
+		"param-val1": appGatewayName,
+		"param-key2": "applicationGatewayWebApplicationFirewallConfiguration",
+		"param-val2": subResname,
+	}
+	subResARMID, err := getFrontendPortsARMID(tc, rg, armid_params)
+	tc.Expect(err).To(BeNil())
+	return subRes, subResARMID
+}
+
+func defineApplicationGatewayFrontendIPConfiguration(tc *testcommon.KubePerTestContext, rg *resources.ResourceGroup, appGatewayName string, subnet *genruntime.ResourceReference, publicIP *genruntime.ResourceReference) ([]network.ApplicationGatewayFrontendIPConfiguration, string) {
+	appGtwFeIpName := "app-gw-fip-config-1"
+	appGtwFeIpName2 := "app-gw-fip-config-2"
+	appGtwFeIpConfig := []network.ApplicationGatewayFrontendIPConfiguration{
+		{
+			Name:                      to.Ptr[string](appGtwFeIpName),
+			PrivateIPAddress:          to.Ptr[string]("10.0.0.10"),
+			PrivateIPAllocationMethod: to.Ptr[network.IPAllocationMethod](network.IPAllocationMethod("Static")),
+			Subnet: &network.ApplicationGatewaySubResource{
+				Reference: subnet,
+			},
+		},
+		{
+			Name: to.Ptr[string](appGtwFeIpName2),
+			PublicIPAddress: &network.ApplicationGatewaySubResource{
+				Reference: publicIP,
+			},
+		},
+	}
+	armid_params := map[string]string{
+		"param-key1": "applicationGateways",
+		"param-val1": appGatewayName,
+		"param-key2": "frontendIPConfigurations",
+		"param-val2": appGtwFeIpName,
+	}
+	appGtwARMID, err := getFrontendPortsARMID(tc, rg, armid_params)
+	tc.Expect(err).To(BeNil())
+	return appGtwFeIpConfig, appGtwARMID
+}
+
+func defineApplicationGatewayFrontendPort(tc *testcommon.KubePerTestContext, rg *resources.ResourceGroup, appGatewayName string) ([]network.ApplicationGatewayFrontendPort, string) {
+	appGtwFePortName := "app-gw-fe-port-1"
+	AppGtwFEPorts := []network.ApplicationGatewayFrontendPort{
+		{
+			Name: to.Ptr[string](appGtwFePortName),
+			Port: to.Ptr[int](443),
+		},
+	}
+	armid_params := map[string]string{
+		"param-key1": "applicationGateways",
+		"param-val1": appGatewayName,
+		"param-key2": "frontendPorts",
+		"param-val2": appGtwFePortName,
+	}
+	appGtwFePortsID, err := getFrontendPortsARMID(tc, rg, armid_params)
+	tc.Expect(err).To(BeNil())
+	return AppGtwFEPorts, appGtwFePortsID
+}
+
+func defineApplicationGatewaySslCertificate(tc *testcommon.KubePerTestContext, rg *resources.ResourceGroup, appGatewayName string) ([]network.ApplicationGatewaySslCertificate, string) {
+	subResname := "app-gw-sslcert-1"
+	subRes := []network.ApplicationGatewaySslCertificate{
+		{
+			//KeyVaultSecretId: to.Ptr[string]("https://keyvaultname.vault.azure.net/secrets/secretname"),
+			Name: to.Ptr[string](subResname),
+			Data: to.Ptr[string](`MIIKEQIBAzCCCdcGCSqGSIb3DQEHAaCCCcgEggnEMIIJwDCCBHcGCSqGSIb3DQEH
```

Comment:
Is this actually going to pass in our CI subscription? Do we need some magical KeyVault with a secret uploaded to it?

Ideally the test would be self-contained, so it would create (and then tear-down) any resources it needed, including the KV and the secret payload. If it's not doing that we need to think about it.

Reply:
If this is an embedded certificate, what's its expiry? 

The test should either be self-contained (creating its own certificate, garanteeing that it's not stale), or at the very least there should be a comment here indicating when the certificate will expire so that a future dev knows what needs doing when the test fails.